fabric-api-base/src/main/java/net/fabricmc/fabric/api/util/TriState.java

```diff
@@ -16,16 +16,197 @@
 
 package net.fabricmc.fabric.api.util;
 
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.BooleanSupplier;
+import java.util.function.Supplier;
+
+/**
+ * Represents a boolean value which can be true, false or refer to a default value.
+ */
 public enum TriState {
+	/**
+	 * Represents the boolean value of {@code false}.
+	 */
 	FALSE,
+	/**
+	 * Represents a value that refers to a "default" value, often as a fallback.
+	 */
 	DEFAULT,
+	/**
+	 * Represents the boolean value of {@code true}.
+	 */
 	TRUE;
 
-	public static TriState of(boolean b) {
-		return b ? TRUE : FALSE;
+	/**
+	 * Gets the corresponding tri-state from a boolean value.
+	 *
+	 * @param bool the boolean value
+	 * @return {@link TriState#TRUE} or {@link TriState#FALSE} depending on the value of the boolean.
+	 */
+	public static TriState of(boolean bool) {
+		return bool ? TRUE : FALSE;
+	}
+
+	/**
+	 * Gets a tri-state from a nullable boxed boolean.
+	 *
+	 * @param bool the boolean value
+	 * @return {@link TriState#DEFAULT} if {@code null}.
+	 * Otherwise {@link TriState#TRUE} or {@link TriState#FALSE} depending on the value of the boolean.
+	 */
+	public static TriState of(/* @Nullable */ Boolean bool) {
+		if (bool == null) {
+			return DEFAULT;
+		}
+
+		return TriState.of(bool.booleanValue());
 	}
 
+	/**
+	 * Gets the value of the tri-state.
+	 *
+	 * @return true if the tri-state is {@link TriState#TRUE}.
+	 * Otherwise false.
+	 */
 	public boolean get() {
 		return this == TRUE;
 	}
+
+	/**
+	 * Gets the value of the tri-state as a boxed, nullable boolean.
+	 *
+	 * @return {@code null} if {@link TriState#DEFAULT}.
+	 * Otherwise {@code true} if {@link TriState#TRUE} or {@code false} if {@link TriState#FALSE}.
+	 */
+	/* @Nullable */
```

Comment:
are you annotating the method or its return type here? imo should be the return type

Reply:
This refers to the return type of the method, the `Boolean`.

Obviously pending on loom 0.5 for jetbrains annotations to make this concretely clear.

Reply:
ah -- I don't know if you all have figured out the code style for that yet, but normally for type annotations i'd stick them directly against the return type (since they're effectively type use annotations), so a method with type and method annotations would look something like:

```java
@Override
public @Nullable Boolean blah() {
    return /* blah */;
}
```

Reply:
Aren't those two annotation styles effectively the same?

Pretty sure there is probably a way to make checkstyle enforce the annotation position in that case.

Reply:
for the compiler, yes -- but for the reader imo that's more clear. Of course this is entirely a code style thing, but hopefully i have *permission* to bikeshed a little

Reply:
If you did want that change, you may as well open an RFC for style change (I want religious use of `this` in codebase so I'll need to open one soon).